File: medium/0102_binary_tree_level_order_traversal/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <queue>
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>> levelOrder(utils::TreeNode *root) {
    if (!root)
      return {};

    std::deque<utils::TreeNode *> dq{root};
    std::vector<std::vector<int>> levels;

    while (!dq.empty()) {
      std::vector<int> level;

      for (size_t _ = dq.size(); _ > 0; --_) {
        utils::TreeNode *node = dq.front();
        dq.pop_front();
        level.push_back(node->val);

        if (node->left)
          dq.push_back(node->left);
        if (node->right)
          dq.push_back(node->right);
      }

      levels.push_back(level);
    }

    return levels;
  }
};

File: medium/0102_binary_tree_level_order_traversal/solution.py
from collections import deque

from utils.python.nodes import TreeNode


class Solution:
    def levelOrder(self, root: TreeNode | None) -> list[list[int]]:
        if not root:
            return []

        queue = deque([root])
        levels = []

        while queue:
            levels.append([])

            for _ in range(len(queue)):
                node = queue.popleft()
                levels[-1].append(node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

        return levels
